Add option to retry on hosted form validation request returning HTTP 5xx

horgh opened the pull request:
This adds two new attributes to `SecureToken`.One to set how many times we will retry if we see a HTTP 5xx response. The second is a callback we call prior to retrying. This callback is useful for logging.

horgh committed:
Remove README

There is a README.md. README appeared very out of date.

horgh committed:
Provide an option to retry hosted form validation requests

The default behaviour is unchanged.

We want to be able to handle transient failures more gracefully by
opting to immediately retry the failed request.

oschwald commented on Changes:
```diff
@@ -2,6 +2,11 @@ Revision history for WebService-PayPal-PaymentsAdvanced
 
 {{$NEXT}}
 
+0.000022 2017-05-08
```
This line is unnecessary. It is automatically generated by dzil on release. We should probably do a release when this PR is merged.

oschwald commented on lib/WebService/PayPal/PaymentsAdvanced.pm:
```diff
@@ -601,6 +604,11 @@ to C<true>.
 L<WebService::PayPal::PaymentsAdvanced::Response::SecureToken> object.
 
+The first parameter holds the key/value parameters for the request. The second
+parameter is optional, and holds parameters to the underlying
```
No comma after "optional".

oschwald commented on lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm:
```diff
@@ -82,6 +100,56 @@ sub _build_hosted_form_uri {
+        my $attempts_left = $self->retry_attempts + 1;
+        $attempts_left > 0;
+        $attempts_left--
+        ) {
```
Any reason to not just do `for my $attempt (1..$self->retry_attempts+1)` here and adjust the `last` check appropriately? The C-style for loop is a bit harder to parse and more prone to off-by-one errors.

horgh replied:
The rationale for the C loop was for the `last` check to be clearly about whether this is the last attempt we have left. However I agree that the count down required more thought. In fact I had an off by one error on my first pass writing it! I've updated it now.

oschwald commented on lib/WebService/PayPal/PaymentsAdvanced/Response/SecureToken.pm:
```diff
@@ -109,6 +177,17 @@ The URL for the PayflowLink web service.  Can be a mocked URL.
+
+The callback to call if we encounter an HTTP 5xx response. We call this prior to
+retrying the request. The callback will have a single parameter, an
+HTTP::Response object.
```
From this, it isn't clear what I am supposed to do with the callback or what its purpose is. A sentence about the motivation or with a use case would go a long way in helping the reader understand.

oschwald commented on t/PaymentsAdvanced/Response/SecureToken.t:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env perl;
```
This should be needed. The `perlcriticrc` file needed to be updated, but now that you rebased against my changes, it should work fine without.

oschwald reviewed (changes requested):
Looks good! I had a couple of minor comments.

horgh committed:
Remove Changes version/date

This gets autogenerated

horgh committed:
Remove unnecessary comma

horgh committed:
Clarify the rationale for the retry_callback

Also adjust the request loop style for clarity, and remove an
unnecessary hashbang.

horgh committed:
Fix typo in comment
